### PR TITLE
add song to playlist implementation

### PR DESCRIPTION
**Pull request by ryachen01**
Issue #258

### CODE REVIEW

File: main/com.spotify/manifest.tt [hunk 1]
```diff
@@ -271,6 +271,18 @@ class @com.spotify
...
+			      in req playlist: Entity(com.spotify:playlist))
+  #_[canonical=["add a song to a playlist", "add a track to a playlist"]]
+  #_[confirmation="add the song $song to the playlist $playlist"]
+  #_[on_error={
+    disallowed_action=["you can't add songs to this playlist"],
+    non_premium_account=["you need a spotify premium account to do this"],
+    rate_limit_error=["spotify is temporarily unavailable"]
+  }]
+  #[confirm=enum auto]
+  #[minimal_projection=[]];
```
Comment (gcampax) [inline main/com.spotify/manifest.tt:284]: Do we need to add entries to dataset.tt for this action as well?

File: main/com.spotify/manifest.tt [hunk 2]
```diff
@@ -271,6 +271,18 @@ class @com.spotify
   }]
   #[confirm=enum(auto)];
 
+  action add_song_to_playlist(in req song : Entity(com.spotify:song),
+			      in req playlist: Entity(com.spotify:playlist))
```
Comment (gcampax) [inline main/com.spotify/manifest.tt:275]: Do we want a better question than the default "What playlist would you like?"

File: main/com.spotify/index.js [hunk 3]
```diff
@@ -1027,6 +1027,35 @@ module.exports = class SpotifyDevice extends Tp.BaseDevice {
         }
     }
 
+    async do_add_song_to_playlist({
+        song,
+        playlist
+    }) {
+        if (this.state.product !== "premium" && this.state.product !== undefined)
+            throwError("non_premium_account");
```
Comment (gcampax) [inline main/com.spotify/index.js:1035]: Does one need premium for playlists?

**Review by gcampax — COMMENTED** [review]
Looks good

File: main/com.spotify/index.js (hunk 3 above)
Comment (ryachen01) [reply main/com.spotify/index.js:1035]: I checked and turns out you don't. I'll correct this.

File: main/com.spotify/manifest.tt (hunk 2 above)
Comment (ryachen01) [reply main/com.spotify/manifest.tt:275]: I think "What playlist would you like?" works well enough. I can't really think of a question that would work better.

File: main/com.spotify/manifest.tt [hunk 4]
```diff
@@ -283,6 +283,16 @@ class @com.spotify
   #[confirm=enum auto]
   #[minimal_projection=[]];
 
+  action create_playlist(in req playlistName : String)
```
Comment (gcampax) [inline main/com.spotify/manifest.tt:286]: we should probably choose `name` as the parameter name (shorter and more generic, makes it easier for the neural models we adopted recently), and we should write a whole bunch of canonical forms around it.

**Comment by ryachen01** [comment]
@gcampax It seems that the agent is incorrectly interpreting some of the scenarios. Should I comment out the NL inputs for now, or will this issue be resolved?

**Comment by gcampax** [comment]
I need to retrain the model, but yeah switch those scenarios to ThingTalk input for now.

File: main/com.spotify/dataset.tt [hunk 5]
```diff
@@ -43,6 +43,16 @@ dataset @com.spotify language "en" {
                    "play ${p_genre:const} music"]]
     ;
 
+    program (p_song :String, p_playlist :String) := (@com.spotify.song(), id =~ p_song) => @com.spotify.add_song_to_playlist(song=id, playlist=p_playlist)
+    #_[utterances=["add ${p_song:no-undefined} to ${p_playlist:no-undefined}",
+                   "put ${p_song:no-undefined} in ${p_playlist:no-undefined}"]]
```
Comment (gcampax) [inline main/com.spotify/dataset.tt:48]: It is reasonable to say "add the song XXX to some playlist" followed by the agent asking what playlist, so I think the `:no-undefined` clause is unnecessary

File: main/com.spotify/eval/scenarios.txt [hunk 6]
```diff
@@ -256,8 +256,9 @@ A: >> expecting = null
 U: \t $dialogue @org.thingpedia.dialogue.transaction.execute;
 U: now => (@com.spotify.playable(), release_date >= makeDate(2000, 1, 1) && release_date <= makeDate(2000, 1, 1) + 1year) => notify;
 
-A: I (see|have( found)?)(one, it is)? Yellow. It is( a)?( pop)? music by Coldplay( released in July 9, 2000)?\. Would you like to play it on Spotify\?
-A: >> expecting = command
+A: I (see|have( found)?) Yellow and The Real Slim Shady\. Yellow is a music by Coldplay, while The Real Slim Shady Eminem wrote\.
```
Comment (gcampax) [inline main/com.spotify/eval/scenarios.txt:259]: We're doing poetry, I see. File this as a bug in genie-toolkit, the priority of the different parts of speech is messed up if we generate this.

File: main/com.spotify/index.js [hunk 7]
```diff
@@ -1027,6 +1027,72 @@ module.exports = class SpotifyDevice extends Tp.BaseDevice {
...
+        let data = {
+            "uris": [song]
+        };
+        return this.add_uris_to_playlist(playListURI, data);
+
+    }
+
+    async findPlaylist(name) {
+        const searchResults = await this.search(name, "playlist", 1);
+        if (!Object.prototype.hasOwnProperty.call(searchResults, 'playlists') || searchResults.playlists.total === 0) throwError('disallowed_action');
```
Comment (gcampax) [inline main/com.spotify/index.js:1062]: We should have a better error code and error message to indicate that we got the playlist name wrong.

File: main/com.spotify/manifest.tt [hunk 8]
```diff
@@ -271,6 +271,70 @@ class @com.spotify
...
+
+  action add_this_song_to_playlist(in req playlist: String
+                                    #[string_values="com.spotify:playlist"]
+                                    #_[prompt=["what playlist do you want to add to"]]
+                                    #_[canonical={
+                                      default="base",
+                                      base=["name"],
+                                      preposition=["named #", "called #", "titled #"],
+                                      property=["name #", "title #"]
+                                    }])
```
Comment (gcampax) [inline main/com.spotify/manifest.tt:309]: Do we need a separate function?
"add the currently playing song to the playlist X" should be:
```
@com.spotify.get_currently_playing() => @com.spotify.add_song_to_playlist(song=song, playlist="X");
```

File: main/com.spotify/manifest.tt [hunk 9]
```diff
@@ -271,6 +271,70 @@ class @com.spotify
...
+                                #_[canonical={
+                                  default="base",
+                                  base=["name"],
+                                  preposition=["named #", "called #", "titled #"],
+                                  property=["name #", "title #"]
+                                }])
+  #_[canonical=["add a song to a playlist", "add a track to a playlist"]]
+  #_[confirmation="add the song $song to the playlist $playlist"]
+  #_[on_error={
+    disallowed_action=["you can't add songs to this playlist"],
```
Comment (gcampax) [inline main/com.spotify/manifest.tt:295]: Always expand all contractions ("can't" -> "cannot", "don't" -> "do not", ...) in canonical forms and phrases, the system will reintroduce contractions as appropriate.

**Review by gcampax — APPROVED** [review]
Looks good!

**Pull request merged**